workloads/imagenet/imagenet_pytorch/workload.py

```diff
@@ -0,0 +1,235 @@
+"""Resnet workload implemented in PyTorch."""
+import contextlib
+import os
+from collections import OrderedDict
+from typing import Tuple
+
+import torch
+import torch.nn.functional as F
+from torchvision import transforms
+from torchvision.datasets import ImageNet
+from torchvision.datasets.folder import ImageFolder
+
+import spec
+import random_utils as prng
+from workloads.imagenet.workload import ImagenetWorkload
+from workloads.imagenet.imagenet_pytorch.resnet import resnet50
+from workloads.imagenet.imagenet_pytorch.utils import fast_collate, cycle
+
+
+DEVICE='cuda' if torch.cuda.is_available() else 'cpu'
+
+
+class ImagenetWorkload(ImagenetWorkload):
+
+  def __init__(self):
+    super().__init__()
+    self.dataset = 'imagenet2012'
+
+  """ data augmentation settings """
+  @property
+  def scale_ratio_range(self):
+    return (0.08, 1.0)
+
+  @property
+  def aspect_ratio_range(self):
+    return (0.75, 4.0 / 3.0)
+
+  @property
+  def center_crop_size(self):
+    return 224
+
+  @property
+  def resize_size(self):
+    return 256
+
+  @property
+  def num_train_examples(self):
+    return 1281167
+
+  @property
+  def num_eval_examples(self):
+    return 50000
+
+  @property
+  def param_shapes(self):
+    """
+    TODO: return shape tuples from model as a tree
+    """
+    raise NotImplementedError
+
+  def eval_model(
+      self,
+      params: spec.ParameterContainer,
+      model_state: spec.ModelAuxiliaryState,
+      rng: spec.RandomState,
+      data_dir: str):
+    """Run a full evaluation of the model."""
+    data_rng, model_rng = prng.split(rng, 2)
+    eval_batch_size = 128
+    if self._eval_ds is None:
+      self._eval_ds = self._build_dataset(
+          data_rng, 'test', data_dir, batch_size=eval_batch_size)
+
+    total_metrics = {
+        'accuracy': 0.,
+        'loss': 0.,
+    }
+    num_batches = 0
+    for (images, labels) in self._eval_ds:
+      images, labels = self.preprocess_for_eval(images, labels, None, None)
+      logits, _ = self.model_fn(
+          params,
+          images,
+          model_state,
+          spec.ForwardPassMode.EVAL,
+          model_rng,
+          update_batch_norm=False)
+      # TODO(znado): add additional eval metrics?
+      batch_metrics = self._eval_metric(logits, labels)
+      total_metrics = {
+          k: v + batch_metrics[k] for k, v in total_metrics.items()
+      }
+      num_batches += 1
+    # Note (runame): the returned result is not exact because the last batch
+    # has a smaller batch size
+    return {k: float(v / num_batches) for k, v in total_metrics.items()}
```

Comment:
Sure, will fix it.

I noticed that in the MNIST PyTorch workload, it's handled in the [same way](https://github.com/mlcommons/algorithmic-efficiency/blob/main/workloads/mnist/workload.py#L78). However, in that case the eval set size is divisible by the eval batch size without remainder, so it's only wrong when a different batch size is used, for which is not the case anymore.

Reply:
I think the best approach is the following:

Let's put the discussion for `eval_batch_size` on the agenda for a future meeting (I can take care of this). We could couple it with a discussion of what information we would want to track in the `eval_model` (see @znado's comment).

Nevertheless, I think fixing it so that it works in the general case would be great (for both ImageNet and MNIST, how is this done in JAX? Do we have the same problem there?). There is no real downside to it (that I can see). At the very least, we should add an assert check that our computation is correct (e.g. that we don't have a smaller final batch).

Reply:
I have fixed the eval metrics for the case where a remainder batch exists, in the MNIST PyTorch + Jax and the ImageNet PyTorch workloads.

One thing I noticed is that `workload.loss_fn()` does not consistently return the loss with the same reduction method applied, i.e. sometimes the per sample loss is returned (e.g. MNIST Jax workload) and sometimes the sum (MNIST/ImageNet PyTorch workloads) or the mean (ImageNet Jax workload) over the loss of each sample in the batch.

Reply:
Great!

In the rules (https://github.com/mlcommons/algorithmic-efficiency/blob/main/RULES.md#loss-function) we stated that we want to return per sample losses (e.g. 1d array of losses per example in batch).

Reply:
Ah I missed the return type in the code snippet in the rules when I checked. I will change this in all workloads accordingly then.